server/paymentProviders/stripe/virtual-cards.js

```diff
@@ -51,3 +58,173 @@ export const assignCardToCollective = async (cardNumber, expireDate, cvv, collec
 
   return await models.VirtualCard.create(cardData);
 };
+
+export const processTransaction = async (stripeEvent, stripeSignature) => {
+  const stripeTransaction = stripeEvent.data.object;
+
+  const virtualCard = await models.VirtualCard.findOne({
+    where: {
+      id: stripeTransaction.card,
+    },
+    include: [
+      { association: 'collective', required: true },
+      { association: 'host', required: true },
+      { association: 'user' },
+    ],
+  });
+
+  if (!virtualCard) {
+    logger.error('stripe/webhook: could not find VirtualCard', { body: stripeTransaction });
+    return;
+  }
+
+  const host = virtualCard.host;
+  const collective = virtualCard.collective;
+  const [connectedAccount] = await host.getConnectedAccounts({ where: { service: 'stripe' } });
+
+  if (!connectedAccount) {
+    logger.error('stripe/webhook: host is not connected to Stripe', { body: stripeTransaction });
+    return;
+  }
+
+  const stripe = Stripe(connectedAccount.token);
+
+  try {
+    stripe.webhooks.constructEvent(stripeEvent, stripeSignature, connectedAccount.data.stripeEndpointSecret);
+  } catch {
+    logger.error('stripe/webhook: source of event not recognized', { body: stripeTransaction });
+    return;
+  }
+
+  const amount = stripeTransaction.amount;
+
+  if (amount === 0) {
+    return;
+  }
+
+  const existingExpense = await models.Expense.findOne({
+    where: {
+      VirtualCardId: virtualCard.id,
+      data: { token: stripeTransaction.id },
+    },
+  });
+  if (existingExpense) {
+    logger.warn(`Virtual Card charge already reconciled, ignoring it: ${stripeTransaction.id}`);
+    return;
+  }
+
+  const UserId = virtualCard.UserId;
+  const isRefund = amount < 0;
```

Comment:
We know that this is how privacy works. But, is it the same with Stripe for Refunds? If it's not, we could be implementing code that is ultimately not needed or need to be different.

Reply:
Wow good catch, I assume it was the same but it is not ! I will modify it accordingly.